Initial Implementation

adellinekania opened the pull request:
Move the deployment commands for OKD from octopus-vcs-facade: https://github.com/octopusden/octopus-vcs-facade/blob/main/buildSrc/src/main/kotlin/org/octopusden/octopus/service/OcTemplateService.kt

ozonophore commented on ft/src/test/kotlin/org/octopusden/octopus/oc/template/plugins/gradle/OcTemplatePluginTest.kt:
```diff
@@ -0,0 +1,173 @@
+        }
+        assertEquals(0, instance.exitCode)
+        assertThat(projectPath.resolve("build/$DEFAULT_WORK_DIR/rest-service.yaml")).exists()
+        assertThat(
```
If I understand correctly, this test deploys a service with a route, and it’s expected that some simple HTTP client will call the endpoint (e.g., curl http://ft-simple-rest-service.app.host.com). The current test only shows that the environment has been deployed, but it doesn’t actually check whether it works. (Even though it’s not common practice here, don’t be afraid to write Javadoc — it helps to understand the test’s purpose faster and not just that.)

adellinekania replied:
You’re right that the test itself doesn’t explicitly make an HTTP call—but the endpoint check is actually included in the [build.gradle](https://github.com/octopusden/octopus-oc-template-gradle-plugin/blob/init-plugin/ft/src/test/resources/projects/rest-service/build.gradle.kts#L34) of the test project. It uses the plugin to retrieve the route and perform a request to verify the deployed service is responding as expected.

ozonophore replied:
@adellinekania Got it. I believe the test scenario should be described in one place using simple and clear steps.
If some steps are in one place and others are elsewhere, the test becomes confusing and non-obvious.
At the very least, it's worth adding some explanation in a Javadoc.

ozonophore commented on ft/src/test/kotlin/org/octopusden/octopus/oc/template/plugins/gradle/OcTemplatePluginTest.kt:
```diff
@@ -0,0 +1,173 @@
+            "-Pokd-namespace=$DEFAULT_OKD_NAMESPACE",
+            "-Pwork-directory=$DEFAULT_WORK_DIR",
+            "-Pproject-prefix=$DEFAULT_PROJECT_PREFIX",
+            "-Pdocker-registry=$DOCKER_REGISTRY"
```
I think these parameters need to be added to the README.

CamaradeRoman commented on ft/src/main/kotlin/org/octopusden/octopus/oc/template/plugins/gradle/OcTemplatePluginRunner.kt:
```diff
@@ -0,0 +1,65 @@
+
+open class TestGradleDSL {
+    lateinit var testProjectName: String
+    lateinit var templateYamlFileName: String
```
Why do we force configure single template file for a test case?

Will be test cases with some set of templates developed?

CamaradeRoman replied:
Would it be better not to parametrize test case with path to yaml file but "hardcode" it inside test project source?

adellinekania replied:
@CamaradeRoman thanks for the review! I updated the code to include tests on the main project, group all parameters validation on `build.gradle.kts`, and declare yaml file property directly on the test project.

CamaradeRoman commented on ft/build.gradle.kts:
```diff
@@ -0,0 +1,18 @@
+}
+
+group = "org.octopusden.octopus"
+
```
I'd prefer to have tests as a part of common project (to have a single build configuration for build, test and publish).

@pgorbachev is there a reason to create separate FT project?

adellinekania commented on ft/src/main/kotlin/org/octopusden/octopus/oc/template/plugins/gradle/OcTemplatePluginRunner.kt:
```diff
@@ -0,0 +1,65 @@
+    return Pair(ProcessBuilders
+        .newProcessBuilder<ProcessBuilder>(LocalProcessSpec.LOCAL_COMMAND)
+        .envVariables(mapOf(
+            "JAVA_HOME" to System.getProperties().getProperty("java.home"),
```
We set `JAVA_HOME` to ensure the test projects run with the same Java version as the test runner. Without it, Gradle may fall back to the system default `JAVA_HOME`, leading to inconsistent test results—especially when multiple JDKs are installed.

CamaradeRoman commented on build.gradle.kts:
```diff
@@ -0,0 +1,138 @@
+
+val testParameters by lazy {
+    mapOf(
+        "ocTemplateGradlePluginVersion" to getTestParameter("OC_TEMPLATE_GRADLE_PLUGIN_VERSION", "ocTemplateGradlePluginVersion", "1.0-SNAPSHOT"),
```
@adellinekania shouldn't we simply use `gradle project version` here?

CamaradeRoman commented on src/test/kotlin/org/octopusden/octopus/oc/template/plugins/gradle/OcTemplatePluginTest.kt:
```diff
@@ -0,0 +1,170 @@
+        const val DEFAULT_WORK_DIR = "okd"
+        const val DEFAULT_PROJECT_PREFIX = "oc-template-ft"
+        private val DEFAULT_TASKS = arrayOf("clean", "build", "--info", "--stacktrace")
+        private val DEFAULT_OKD_NAMESPACE: String = System.getProperty("okdNamespace")
```
Why is it named DEFAULT_? Do we change it in some test cases?

CamaradeRoman commented on build.gradle.kts:
```diff
@@ -0,0 +1,138 @@
+}
+
+tasks.test {
+    useJUnitPlatform()
```
@adellinekania it is impossible to configure compilation in tests in single run - no dependency on publishToMavenLocal. So locally tests are using some previously "installed" snapshot version. That may lead to "unexpected behaviour" if one forgot to "reinstall" changed snapshot to local repository.

adellinekania commented on src/test/kotlin/org/octopusden/octopus/oc/template/plugins/gradle/runner/OcTemplatePluginRunner.kt:
```diff
@@ -0,0 +1,55 @@
+        .newProcessBuilder<ProcessBuilder>(LocalProcessSpec.LOCAL_COMMAND)
+        .envVariables(mapOf(
+            "JAVA_HOME" to System.getProperty("java.home"),
+            "OKD_CLUSTER_DOMAIN" to System.getProperty("okdClusterDomain")
```
I’ve moved it to `OcTemplatePluginTest`. Initially, I had placed it in `OcTemplatePluginRunner` because it wasn’t being overridden by anything, but I agree that it makes more sense to have it in the test layer. Thanks for the suggestion!

CamaradeRoman commented on build.gradle.kts:
```diff
@@ -0,0 +1,139 @@
+    `maven-publish`
+    signing
+    id("io.github.gradle-nexus.publish-plugin")
+    id("com.jfrog.artifactory")
```
Why do we require `com.jfrog.artifactory` plugin? We don't use it in any other project.

adellinekania commented on src/test/kotlin/org/octopusden/octopus/oc/template/plugins/gradle/OcTemplatePluginTest.kt:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    private fun getLogFileName(serviceName: String): String {
+        return "$DEPLOYMENT_PREFIX-1-0-snapshot-$serviceName.log"
```
It works on TeamCity with the specified version because the `1-0-snapshot` here refers to the test project’s version, not the plugin’s version.

adellinekania commented:
@CamaradeRoman I’ve added logic for version calculation: if the project version is not specified, the plugin will now generate a `-snapshot` version. Let me know if that works for you.

CamaradeRoman commented on build.gradle.kts:
```diff
@@ -0,0 +1,139 @@
+            contextUrl = "$baseUrl/artifactory"
+        }
+        repository {
+            repoKey = System.getenv().getOrDefault("ARTIFACTORY_REPOSITORY_KEY", project.properties["artifactoryRepositoryKey"]).toString()
```
@adellinekania I expect here hardcoded value like in
https://github.com/octopusden/octopus-rm-gradle-plugin/blob/f52a8c4b41069ad13953b0768014394fb8600052/build.gradle#L91
because we specify it here https://github.com/octopusden/octopus-rm-gradle-plugin/blob/f52a8c4b41069ad13953b0768014394fb8600052/src/main/groovy/org/octopusden/release/management/plugins/gradle/ReleaseManagementGradlePlugin.groovy#L199

adellinekania commented:
@CamaradeRoman, I’ve updated the code to use `generateDefaultVersion()` when the version is unspecified or ends with `-snapshot`. Additionally, I added logging for the pod URL after the readiness check.
